mux_client_impl/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+//! remote end. That shell _interprets_ anything passed to it — it might evalute words starting
+//! with `$` as variables, split arguments by whitespace, and other things a shell is wont to do.
+//! Since that is _usually_ not what you expect to happen, `.arg("a b")` should pass a _single_
+//! argument with the value `a b`, `openssh` _escapes_ every argument (and the command itself) by
+//! default using [`shell-escape`]. This works well in most cases, but might run into issues when
+//! the remote shell (generally the remote user's login shell) has a different syntax than the
+//! shell `shell-escape` targets (bash). For example, Windows shells have different escaping syntax
+//! than bash does.
+//!
+//! If this applies to you, you can use [`raw_arg`](Command::raw_arg),
+//! [`raw_args`](Command::raw_args), and [`raw_command`](Session::raw_command) to bypass the
+//! escaping that `openssh` normally does for you.
+//!
+//! # Examples
+//!
+//! ```rust,no_run
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), mux_client_impl::Error> {
+//! use mux_client_impl::{Session, KnownHosts};
+//!
+//! let session = Session::connect("me@ssh.example.com", KnownHosts::Strict).await?;
+//! let ls = session.command("ls").output().await?;
+//! eprintln!("{}", String::from_utf8(ls.stdout).expect("server output was not valid UTF-8"));
+//!
+//! let whoami = session.command("whoami").output().await?;
+//! assert_eq!(whoami.stdout, b"me\n");
+//!
+//! session.close().await?;
+//! # Ok(()) }
+//! ```
+//!
+//!   [`ControlMaster`]: https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Multiplexing
+//!   [`sshd_config`]: https://linux.die.net/man/5/sshd_config
+//!   [`shell-escape`]: https://crates.io/crates/shell-escape
+
+#![warn(
+    missing_docs,
+    missing_debug_implementations,
+    rustdoc::broken_intra_doc_links,
+    rust_2018_idioms,
+    unreachable_pub
+)]
+
+use std::borrow::Cow;
+use std::path;
+
+use shell_escape::escape;
+use tempfile::TempDir;
+
+use openssh_mux_client::connection::Connection;
+pub use openssh_mux_client::connection::{ForwardType, Socket};
+
+mod builder;
+pub use builder::{KnownHosts, SessionBuilder};
+
+mod fd;
+pub(crate) use fd::*;
+
+mod stdio;
+pub use stdio::{ChildStderr, ChildStdin, ChildStdout, Stdio};
+
+mod command;
+pub use command::Command;
+
+mod child;
+pub use child::RemoteChild;
+
+mod error;
+pub use error::Error;
+/// Typedef just like std::io::Error
+pub type Result<T, Err = Error> = std::result::Result<T, Err>;
+
+#[cfg(test)]
+#[macro_use]
+extern crate assert_matches;
+
+/// A single SSH session to a remote host.
+///
+/// You can use [`command`] to start a new command on the connected machine.
+///
+/// When the `Session` is dropped, the connection to the remote host is severed, and any errors
+/// silently ignored. To disconnect and be alerted to errors, use [`close`](Session::close).
+#[derive(Debug)]
+pub struct Session {
```

Comment:
Here, too, I'd like to just have _one_ top-level `Session` type that the user interacts with, and then have it just be generic over the backing implementation.